loop streamlining

patricksurry opened the pull request:
Here are a few ideas for do loop optimization while I was figuring out how it worked.

The ROM size is +6 bytes where some space optimizations were compensated by adding a specialized loop_runtime (+1 instead of +step).

The runtime is about 40 bytes less for each instance by making do_runtime a subroutine, so slightly more expensive on loop startup, but the loop step (for step=1) is shorter and quicker (avoiding the jsr xt_one stack juggling stuff).  

Saved a couple of bytes in xt_i but wondering if there's something better there.   If you assume i is accessed in most loops it might be better to speculatively increment the index at the same time as incremental the fudged limit in the loop footer?

patricksurry committed:
loop streamlining

SamCoVT commented:
Having a special runtime for +1 looping seems reasonable.  In my own Forth code, I'm much more likely to be either counting or doing sequential things, so I use `LOOP` way more often than `+LOOP`.

Sometimes I use `I` in loops and sometimes I don't, so I'm not sure if it's worth having the fudge factor or not.  It's designed to make it so the sign bit in the MSB can be used to know when to stop (note that with `+LOOP`, it's possible that you don't stop _exactly on_ the ending value, so you have to handle passing the ending value).  It's a hassle to "un-fudge" the index for `I`, though.

The testing for loops is reasonably thorough, so you'll likely know quickly if you break anything.  There currently isn't any cycle counting testing for do loops.  Perhaps it makes sense to define some words with various kinds of do loops and cycle count them?  It would be good to have a mix of w/ and w/out `I`, nested loops, and using `+LOOP` with both positive and negative values.

Some other Forths keep the top-level loop index/limit in zero page and only push them onto the return stack when starting another loop inside a loop.  This makes for much faster access to the index.  `I` is more likely to be used than `J`, and `J` wouldn't really be any slower than it is now.  Tali doesn't have much zero page free (due to a self-imposed limit of only using half of zero page), but it might be possible to make some room.

Some other Forths have a control stack, separate from the data stack and return stack.  I'm not sure that helps us here.

patricksurry commented on native_words.asm:
```diff
@@ -5893,8 +5863,7 @@ plus_loop_runtime:
 
-                tya             ; put LSB of index back on stack
-                pha
+                phy             ; put LSB of index back on stack
```
-1 byte with 65c02 inst

SamCoVT replied:
There are probably lots more small 65C02 specific speedups in Tali, so if you see any others while looking through the source code, I'd be happy for a pull request even for something as small as this.

patricksurry commented on native_words.asm:
```diff
@@ -5843,39 +5818,34 @@ xt_plus_loop:
                 sec
                 sbc #1
-                sta tmp2
+                pha             ; lsb
```
push saves a couple of bytes and simplifies template copy below

SamCoVT replied:
This looks good.  Using the tmpx variables needs a LOT of care to make sure any words that are JSRd to don't also use them (or you push/pop them).  It looks like you've removed some of the tmpx usage, which is great.

patricksurry commented on native_words.asm:
```diff
@@ -5843,39 +5818,34 @@ xt_plus_loop:
-                iny
-                lda tmp2        ; LSB
+-
+                lda loop_epilogue,y
```
copy template, inserting stacked bytes in $0 placeholders

patricksurry commented on native_words.asm:
```diff
@@ -3089,21 +3089,10 @@ do_common:
-                inc CP+1
-+
+                ldy #6
+                jsr cmpl_inline        ; write 6 dummy bytes to runtime
```
we actually don't need the write part at all, just the cp increment, but we have this routine already

patricksurry commented on native_words.asm:
```diff
@@ -3175,6 +3144,12 @@ do_runtime:
         ; However, we can do it faster if we just copy the bytes
         ; of this routine with a simple loop in DO.
         ; """
+                ; stash return address so we can manipulate return stack
```
slightly more expensive to call this as a subroutine, but it only happens on loop startup and saves a bunch of space on every use of DO

SamCoVT replied:
As you already mentioned, the DO/LOOP system is already heavy in both bytes and cycles.  I think this is a good tradeoff.  It might be worth adding this to the debugger dispatch table - alternatively, it could be added to the dictionary as something like `(DO)`.

patricksurry replied:
i'm not sure what you mean when you say debugger dispatch table.  is that so it can appear labeled in disasm or something else?

SamCoVT replied:
Yes, so it will be labeled in the disassembler (which is what I meant when I wrote debugger - my bad) - otherwise it will be an unlabeled JSR and might not be clear that it's the beginning of a loop.

SamCoVT commented on native_words.asm:
```diff
@@ -5843,39 +5818,34 @@ xt_plus_loop:
                 iny
-                lda #$48        ; Opcode for PHA
-                sta (tmp1),y
+                cpy #6
```
If you put a label at the end of the epilogue, you can use label math here instead of a magic number.  That might also be handy if you end up fundamentally changing how the loops work and need to edit the epilogue.

SamCoVT commented on native_words.asm:
```diff
@@ -5908,9 +5877,37 @@ _hack:          ; This is why this routine must be natively compiled: We
+                clv             ; note inc doesn't affect V
+                pla             ; LSB of index
+                ina
+                tay             ; stash updated LSB
```
I think pla-ina-tay can be reduced to:
```
    ply
    iny
```
as the section after this reuses A.

SamCoVT commented on taliforth.asm:
```diff
@@ -102,6 +102,25 @@ cmpl_a:
 _done:
                 rts
 
+cmpl_inline:
```
Tali already has a heavy-duty inline compiler (can inline compile any size routine) in native_words.asm (xt_compile_comma).  It might make sense to use that instead of this routine (or pull it out of that word).  

If you want to use that routine where it currently lives, you will need a bit of setup.  In particular, that full routine tries to look up the NT and will compile as a JSR if there isn't one.  You could:

A. Add your routine you want to compile to the dictionary with the AN (Always Native) flag set.  Then you'd just need to place its XT on the stack and call xt_compile_comma.

B. Alternatively, you can just sneak into the _compile_as_code: section of the routine.  I believe you'll need to remove the _ in the label name to make it a globally available label.  This section expects the following:
Beginning address of source code on return stack
Length on data stack
status byte in tmp3 (load with 0 to prevent underflow stripping)

SamCoVT replied:
The _specials_done label might be an even better target.  This needs the following on the stack:
source_address cp length
It calls `move`, and then updates cp.

patricksurry replied:
ok i'll take a look - basically I just saw this code block repeated in three places for compilation so reflexively made a subroutine :)

SamCoVT replied:
That totally makes sense.  The version we already have requires a bit more setup and will run slower, but this is during compiling.  I'm not as concerned with the speed during compiling if it will be faster and/or smaller when running.

patricksurry committed:
PR feedback

patricksurry commented on disassembler.asm:
```diff
@@ -443,6 +444,14 @@ disasm_branch:
+                lda #'O'
+                jsr emit_a
+                rts
+
```
it didn't seem worth creating a whole new string for this but not sure if this is what you intended

SamCoVT replied:
I think that's less bytes than adding a string and putting it into the lookup table, so that's fine.  It is what I was intending.  
It's technically not a JSR to `do` itself, but I don't think we need extra effort to attempt to make that clear.  You did put it in uppercase here, so that makes it a little bit distinct - even through Tali is not case sensitive, if someone did JSR to `do` then Tali would list that in lowercase as the words are lowercased in the dictionary.

patricksurry commented on native_words.asm:
```diff
@@ -2116,7 +2116,7 @@ _compile_check:
                 beq _check_size_limit
 
-                jmp _compile_as_jump    ; too far for BRA
+                jmp compile_as_jump    ; too far for BRA
```
creating a new entry point meant globalizing some other labels.  not sure if there's a better way to do that

SamCoVT replied:
I don't think there is a better way to do that.  I've actually considered making all of Tali's labels global and naming all of the anonymous labels in order to make it easier to switch assemblers, but then I remember how many _done labels there are :)

patricksurry committed:
weird caps

patricksurry commented on native_words.asm:
```diff
@@ -2297,22 +2297,36 @@ _specials_done:
+                rts
 
-_strip_table:
+cmpl_inline_y:      ; with ( src -- ) and Y = # bytes, write to CP
```
this is a helper to do prep shared stack for cmpl_inline

patricksurry commented on native_words.asm:
```diff
@@ -3082,27 +3095,19 @@ do_common:
                 lda cp+1
                 sta 1,x                 ; MSB   ( limit start here )
+                lda cp
+                sta 0,x                 ; LSB
```
swap so I can reuse A, realized we don't actually need to write anything, just advance cp

patricksurry commented:
I think I addressed all of your feedback.  

Here's what `0 nc-limit !  : foo 0 10 1 do i + loop ;  see foo` looks like now.

The `i` word is still pretty heavy (and inlined) esp when increment is 1.  I'm still tempted to also increment `i` in `loop` /  `+loop`, and keep it on the stack instead of the base FUFA.  Then `i` could be a lightweight subroutine (or inline) that just makes a copy.   

I guess also arguable whether the end `loop` / `+loop` should always be inlined - a subroutine is a little more expensive but saves a fair bit of space.  Presumably tricky to make it work both ways since the stack arithmetic would have to shift by two.   Or is there already machinery for that?  


```
size (decimal): 68 

080B  20 83 A7 20 9F 93 0A 00  20 12 98 A9 08 48 A9 4E   .. ....  ....H.N
081B  48 20 B9 8B CA CA DA BA  38 BD 02 01 FD 04 01 A8  H ...... 8.......
082B  BD 03 01 FD 05 01 FA 95  01 94 00 20 ED 99 B8 7A  ........ ... ...z
083B  C8 D0 05 68 18 69 01 48  5A 70 03 4C 1F 08 68 68  ...h.i.H Zp.L..hh
084B  68 68 68 68  hhhh

80B   A783 jsr     0
80E   939F jsr     LITERAL A 
813   9812 jsr     1
816      8 lda.#
818        pha
819     4E lda.#
81B        pha
81C   8BB9 jsr     DO
81F        dex
820        dex
821        phx
822        tsx
823        sec
824    102 lda.x
827    104 sbc.x
82A        tay
82B    103 lda.x
82E    105 sbc.x
831        plx
832      1 sta.zx
834      0 sty.zx
836   99ED jsr     +
839        clv
83A        ply
83B        iny
83C      5 bne      843 v
83E        pla
83F        clc
840      1 adc.#
842        pha
843        phy
844      3 bvs      849 v
846    81F jmp
849        pla
84A        pla
84B        pla
84C        pla
84D        pla
84E        pla
```

SamCoVT commented on native_words.asm:
```diff
@@ -2274,7 +2274,7 @@ _underflow_strip:
 
                 ; --- END OF SPECIAL CASES ---
-_specials_done:
+cmpl_inline:    ; ( src dst u -- )
```
This is a much better name for this section.

SamCoVT commented on native_words.asm:
```diff
@@ -2297,22 +2297,36 @@ _specials_done:
                 sta cp+1
 
-                bra _done
+                rts
```
Just a note on RTS in the middle of a word - you can get away with it here because `COMPILE,` is flagged as a NeverNative word, but most words need to have a single RTS at the very end to allow them to be natively compiled.  That's why you'll see things like `bra _done` all over the place.

SamCoVT commented on native_words.asm:
```diff
@@ -5885,9 +5911,8 @@ loop_runtime:
-                ina
-                tay             ; stash updated LSB
+                ply             ; LSB of index
+                iny             ; stash updated LSB
```
The comment no longer matches the assembly.

patricksurry committed:
fix comment

SamCoVT commented:
> The `i` word is still pretty heavy (and inlined) esp when increment is 1. I'm still tempted to also increment `i` in `loop` / `+loop`, and keep it on the stack instead of the base FUFA. Then `i` could be a lightweight subroutine (or inline) that just makes a copy.
> 
> I guess also arguable whether the end `loop` / `+loop` should always be inlined - a subroutine is a little more expensive but saves a fair bit of space. Presumably tricky to make it work both ways since the stack arithmetic would have to shift by two. Or is there already machinery for that?

I believe the loop words are flagged AlwaysNative, and it's precisely because if they were a JSR the values would be in a different place on the return stack.  It's also to save 12 cycles of a JSR/RTS.  There is currently no easy way to handle both a JSR and natively compiled version.

Moving the current value of `I` into zero page would give a significant speedup and size reduction, as accessing it would be just zero page accesses, but we'd need to find a place to put it.  There are other zero-page system variables that could be moved into the RAM system variables to make room.  I think, ideally, we should have the fudged value and the fudge factor in zero page to speed up I as much as possible.  When starting a loop, the values in those would be pushed onto the stack (so `J` would still be slower) to make room for the current values.

Does it make sense to pull what you have so far?  It is already better/faster.  If you are interested in trying out a zero page version, you should let me know and I can shuffle some things around first to make room for the variables you'll need.

patricksurry commented:
yup, i think this current state is a decent intermediate point to pull.

if you give me some help on organizing the storage (and maybe a bit more explanation of how nested loops should behave in terms of pushing things to stack) i'd be game to try out a zp version next week

patricksurry commented:
see #53
